### PR TITLE
Rabbit hole of docgen changes

### PR DESCRIPTION
**Pull request by superm1**
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [x] Documentation

### CODE REVIEW

**Review by hughsie — APPROVED** [review]

**Comment by superm1** [comment]
>hughsie approved these changes 4 hours ago

OK, don't merge yet, I wanted to make sure directionally you were good with this.  I'll push more then.

File: plugins/dfu/fu-dfu-target-stm.c [hunk 1]
```diff
@@ -6,6 +6,8 @@
 
 #include "config.h"
 
+#include <fwupd.h>
+#include <fwupdplugin.h>
```
Comment (hughsie) [inline plugins/dfu/fu-dfu-target-stm.c:10]: you don't need the former if you have the latter -- there's a few of these here too

Comment (hughsie) [reply plugins/dfu/fu-dfu-target-stm.c:10]: I've added a precommit script to detect this; feel free to roll it in with yours. Thanks for all the docs work; it's looking great.

**Comment by superm1** [comment]
@hughsie In the process I noticed a lot of plugins using private headers.  This is fine for the plugin test suites, but shouldn't go beyond that IMO.  So I modified the check-headers.py you added to catch them.
I think we need to make decisions about all of these whether to promote symbols into ABI, change, or waive. 

```
plugins/dfu/fu-dfu-target.c uses private header fu-dfu-target-private.h
plugins/dfu/fu-dfu-target.c uses private header fu-dfu-firmware-private.h
plugins/dfu/fu-dfu-device.c uses private header fu-dfu-target-private.h
plugins/dfu/fu-dfu-device.c uses private header fu-dfu-firmware-private.h
plugins/dfu/fu-dfu-target-avr.c uses private header fu-dfu-target-private.h
plugins/dfu/fu-dfu-target-stm.c uses private header fu-dfu-target-private.h
plugins/dfu/fu-dfu-tool.c uses private header fu-context-private.h
plugins/thunderbolt/fu-thunderbolt-device.c uses private header fu-udev-device-private.h
plugins/modem-manager/fu-mm-device.c uses private header fu-device-private.h
plugins/tpm-eventlog/fu-tpm-eventlog.c uses private header fwupd-common-private.h
plugins/uefi-capsule/fu-uefi-tool.c uses private header fu-context-private.h
```
MM I opened https://github.com/fwupd/fwupd/pull/3348
TBT I pushed a fixup
uefi-tool and dfu-tool I think it's ok to waive

So basically dfu and tpm-eventlog left to figure out what to do.

**Comment by hughsie** [comment]
I've pushed some of this to master, and moved the simple bits to https://github.com/fwupd/fwupd/pull/3349 -- lets commit the latter (assuming CI passes) and then rebase this PR on top of the new master -- this PR is excellent stuff, but is also is getting a bit unwieldy.

**Comment by hughsie** [comment]
DFU bits in https://github.com/fwupd/fwupd/pull/3350

**Comment by superm1** [comment]
Thanks for the split and merge of various things! I rebased it on master and so it's just the bits about what to do with plugins still referencing `*-private.h` stuff.  

BTW, It looks like documentation published properly for the new gidocgen stuff 😀

File: contrib/ci/check-headers.py [hunk 2]
```diff
@@ -31,16 +34,40 @@ def test_files() -> int:
...
+    lib_headers = glob.glob('libfwupd*/*.h')
+    lib_headers.remove ('libfwupd/fwupd.h')
+    lib_headers.remove ('libfwupdplugin/fwupdplugin.h')
+    lib_headers_nopath = [os.path.basename(fn) for fn in lib_headers]
+
+    # test all C and H files
+    for fn in glob.glob("**/*.[c|h]", recursive=True):
+        includes = __get_includes(fn)
+
+        if fn.startswith('plugins'):
```
Comment (hughsie) [inline contrib/ci/check-headers.py:46]: maybe something like `and not fn.endswith('self-test.c')` would reduce the amount of annotations required?

File: contrib/ci/check-headers.py [hunk 3]
```diff
@@ -17,6 +18,8 @@ def __get_includes(fn: str) -> List[str]:
         for line in f.read().split("\n"):
             if line.find("#include") == -1:
                 continue
+            if line.endswith("//waive-pre-commit"):
```
Comment (hughsie) [inline contrib/ci/check-headers.py:21]: `line.find("waive-pre-commit") == -1` might be more flexible.

**Comment by hughsie** [comment]
>  It looks like documentation published properly for the new gidocgen stuff 

Dude, it looks epic.

**Comment by superm1** [comment]
> `!= -1`, no?

No, that will cause it to skip every single include.

**Comment by superm1** [comment]
> maybe ignore -tool too?

maybe, I guess so.

File: contrib/ci/check-headers.py [hunk 4]
```diff
@@ -31,18 +34,46 @@ def test_files() -> int:
...
-        for fn in glob.glob("**/*.[c|h]", recursive=True):
-            includes = __get_includes(fn)
+            toplevel_fn = os.path.basename(toplevel_header)
+            toplevel_includes = __get_includes(toplevel_header)
+            toplevel_includes_nopath = [
+                os.path.basename(fn) for fn in toplevel_includes
+            ]
 
-            # we do not need both toplevel headers
+            # we do not need both toplevel haders
```
Comment (hughsie) [inline contrib/ci/check-headers.py:76]: you reverted my typo fix :)

File: plugins/dfu/fu-dfu-device.c [hunk 5]
```diff
@@ -75,7 +75,7 @@
 #include "fu-dfu-common.h"
 #include "fu-dfu-device.h"
 #include "fu-dfu-target-avr.h"
-#include "fu-dfu-target-private.h"
+#include "fu-dfu-target-private.h" //waive-pre-commit
```
Comment (hughsie) [inline plugins/dfu/fu-dfu-device.c:78]: it makes me a bit uneasy seeing C++ comments for this -- we've used /**/ literally everywhere else in the codebase.

File: contrib/ci/check-headers.py (hunk 4 above)
Comment (superm1) [reply contrib/ci/check-headers.py:76]: Yeah, I missed it when rebasing.
Fixed it.

File: libfwupd/fwupd-common.h [hunk 6]
```diff
@@ -91,5 +91,6 @@ gchar		*fwupd_guid_hash_string			(const gchar	*str);
 gchar		*fwupd_guid_hash_data			(const guint8	*data,
 							 gsize		 datasz,
 							 FwupdGuidFlags	 flags);
+gchar		*fwupd_checksum_format_for_display	(const gchar	*checksum);
```
Comment (hughsie) [inline libfwupd/fwupd-common.h:94]: I think drop this, i merged it to master with the prototype in the right place.

**Review by hughsie — DISMISSED** [review]
One NIT, then LGTM.
